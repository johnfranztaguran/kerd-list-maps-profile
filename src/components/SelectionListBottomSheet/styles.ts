import { StyleSheet, ViewStyle } from 'react-native'

interface Styles {
  selectionBottomListHeaderWrapper: ViewStyle
  selectionListOptionWrapper: ViewStyle
}

const styles: Styles = StyleSheet.create<Styles>({
  selectionBottomListHeaderWrapper: {
    zIndex: 2,
    overflow: 'visible'
  },
  selectionListOptionWrapper: {
    padding: 2
  }
})

export default styles
