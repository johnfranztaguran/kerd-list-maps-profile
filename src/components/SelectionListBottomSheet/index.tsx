import React, { useState } from 'react'
import { Platform, TouchableOpacity, FlatList } from 'react-native'
import { window } from 'constant'
import BottomSheet, { BottomSheetProps } from '../BottomSheet'
import styles from './styles'

export type SelectionListBottomSheetProps<T> = BottomSheetProps & {
  items: Array<T>
  renderListItem: (item: T) => JSX.Element
  onSelectItem: (item: T) => void
  closeOnSelect?: boolean
  height?: number
}

const getValidPositionWithinScrollLimits = (
  val: number,
  min: number,
  max: number
): number => {
  if (val > max) {
    return max
  }

  if (val < min) {
    return min
  }

  return val
}

const SelectionListBottomSheet = <T extends any>(
  props: SelectionListBottomSheetProps<T>
) => {
  const maxOffset: number = 30

  const {
    headerTitle = '',
    bottomSheetRef,
    items = [],
    renderListItem,
    onSelectItem,
    onClose,
    closeOnSelect = true,
    height
  } = props

  const onCloseBottomSheet = () => {
    setOpacity(0)

    if (onClose) {
      onClose()
    }
  }

  const onPressItem = (item: T) => {
    onSelectItem(item)

    if (closeOnSelect) {
      bottomSheetRef?.current?.close()
      onCloseBottomSheet()
    }
  }

  const [opacity, setOpacity] = useState(0)

  const updateOffset = (val: number) => {
    setOpacity(
      getValidPositionWithinScrollLimits((val * maxOffset) / 100, 0, 1)
    )
  }

  return (
    <BottomSheet
      headerTitle={headerTitle}
      bottomSheetRef={bottomSheetRef}
      onClose={onCloseBottomSheet}
      height={height ?? window.height / 2}
      headerStyle={[
        styles.selectionBottomListHeaderWrapper,
        {
          ...Platform.select({
            ios: {
              shadowOpacity: opacity ?? 0
            },
            android: {
              elevation: opacity > 0 ? 10 : 0
            }
          })
        }
      ]}>
      <FlatList<T>
        onScroll={event => updateOffset(event.nativeEvent.contentOffset.y)}
        keyExtractor={item => JSON.stringify(item)}
        showsVerticalScrollIndicator={false}
        data={items}
        renderItem={({ item }) => (
          <TouchableOpacity
            style={styles.selectionListOptionWrapper}
            onPress={() => onPressItem(item)}>
            {renderListItem(item)}
          </TouchableOpacity>
        )}
      />
    </BottomSheet>
  )
}

export default SelectionListBottomSheet
