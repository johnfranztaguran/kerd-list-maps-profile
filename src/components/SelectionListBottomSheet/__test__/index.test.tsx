import React from 'react'
import { View as MockView, Text as MockText } from 'react-native'
import { render } from '@testing-library/react-native'
import SelectionListBottomSheet from '../index'

jest.mock('react-native-gesture-handler', () => {
  const View = require('react-native/Libraries/Components/View/View')
  return {
    Swipeable: View,
    DrawerLayout: View,
    State: {},
    ScrollView: View,
    Slider: View,
    Switch: View,
    TextInput: View,
    ToolbarAndroid: View,
    ViewPagerAndroid: View,
    DrawerLayoutAndroid: View,
    WebView: View,
    NativeViewGestureHandler: View,
    TapGestureHandler: View,
    FlingGestureHandler: View,
    ForceTouchGestureHandler: View,
    LongPressGestureHandler: View,
    PanGestureHandler: View,
    PinchGestureHandler: View,
    RotationGestureHandler: View,
    /* Buttons */
    RawButton: View,
    BaseButton: View,
    RectButton: View,
    BorderlessButton: View,
    /* Other */
    FlatList: View,
    gestureHandlerRootHOC: jest.fn(),
    Directions: {}
  }
})

describe('Component: SelectionListBottomSheet', () => {
  it('Should render a SelectionListBottomSheet', () => {
    const { baseElement } = render(
      <SelectionListBottomSheet<string>
        items={['a', 'b', 'c']}
        renderListItem={item => (
          <MockView>
            <MockText>{item}</MockText>
          </MockView>
        )}
        onSelectItem={jest.fn}
      />
    )
    expect(baseElement).toMatchSnapshot()
  })

  it('Should render a SelectionListBottomSheet with height set', () => {
    const { baseElement } = render(
      <SelectionListBottomSheet<string>
        items={['a', 'b', 'c']}
        renderListItem={item => (
          <MockView>
            <MockText>{item}</MockText>
          </MockView>
        )}
        onSelectItem={jest.fn}
        height={100}
      />
    )
    expect(baseElement).toMatchSnapshot()
  })
})
