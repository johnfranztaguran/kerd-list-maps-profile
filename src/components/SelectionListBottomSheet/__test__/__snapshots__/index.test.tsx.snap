// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: SelectionListBottomSheet Should render a SelectionListBottomSheet 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <Modal
      animationType="none"
      hardwareAccelerated={false}
      supportedOrientations={
        Array [
          "portrait",
          "portrait-upside-down",
          "landscape",
          "landscape-left",
          "landscape-right",
        ]
      }
      transparent={true}
      visible={false}
    >
      <View
        style={
          Array [
            Array [
              Object {
                "backgroundColor": "#00000077",
                "flex": 1,
              },
              undefined,
            ],
            Object {
              "paddingBottom": 0,
            },
          ]
        }
      >
        <TouchableOpacity
          activeOpacity={1}
          style={
            Object {
              "backgroundColor": "transparent",
              "flex": 1,
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "#fff",
              "borderTopLeftRadius": 20,
              "borderTopRightRadius": 20,
              "height": 0,
              "overflow": "hidden",
              "transform": Array [
                Object {
                  "translateX": 0,
                },
                Object {
                  "translateY": 0,
                },
              ],
              "width": "100%",
            }
          }
        >
          <SafeAreaView
            style={
              Object {
                "flex": 1,
              }
            }
          >
            <View
              accessibilityLabel="view-"
              accessible={true}
              style={
                Array [
                  Object {
                    "backgroundColor": "#FFFFFF",
                    "shadowColor": "rgba(0, 0, 0, 0.12)",
                    "shadowOffset": Object {
                      "height": 1,
                      "width": 0,
                    },
                    "shadowRadius": 5,
                  },
                  Object {
                    "shadowOpacity": 0,
                  },
                  Array [
                    Object {
                      "overflow": "visible",
                      "zIndex": 2,
                    },
                    Object {
                      "shadowOpacity": 0,
                    },
                  ],
                ]
              }
              testID="view-"
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "paddingHorizontal": 16,
                    "paddingTop": 12,
                  }
                }
              >
                <View
                  style={
                    Object {
                      "flex": 1,
                    }
                  }
                >
                  <Text
                    accessibilityLabel="txt-"
                    accessible={true}
                    style={
                      Array [
                        Object {
                          "color": "#000000",
                          "fontFamily": "System",
                          "fontWeight": "400",
                          "textAlign": "left",
                        },
                        Array [
                          Object {
                            "color": "#000000",
                            "fontSize": 12,
                          },
                          undefined,
                          undefined,
                          undefined,
                          undefined,
                          Object {
                            "fontFamily": "OpenSans",
                            "fontSize": 16,
                            "lineHeight": 19,
                          },
                          undefined,
                          undefined,
                          undefined,
                          undefined,
                          false,
                          undefined,
                          undefined,
                          false,
                          false,
                          undefined,
                          undefined,
                          undefined,
                          undefined,
                          undefined,
                          undefined,
                          undefined,
                          undefined,
                          false,
                          undefined,
                          undefined,
                          undefined,
                          undefined,
                          undefined,
                          undefined,
                          undefined,
                        ],
                      ]
                    }
                    testID="txt-"
                  >
                    
                  </Text>
                </View>
                <TouchableOpacity
                  accessibilityLabel="btn-close-button"
                  accessible={true}
                  activeOpacity={0.2}
                  style={
                    Object {
                      "alignItems": "flex-end",
                      "height": 44,
                      "justifyContent": "center",
                      "marginRight": 0,
                      "width": 44,
                    }
                  }
                  testID="btn-close-button"
                >
                  <Text
                    allowFontScaling={false}
                    style={
                      Array [
                        Object {
                          "color": undefined,
                          "fontSize": 24,
                        },
                        undefined,
                        Object {
                          "fontFamily": "Material Icons",
                          "fontStyle": "normal",
                          "fontWeight": "normal",
                        },
                        Object {},
                      ]
                    }
                  >
                    Óóç
                  </Text>
                </TouchableOpacity>
              </View>
            </View>
            <View
              style={
                Object {
                  "flex": 1,
                  "paddingHorizontal": 16,
                }
              }
            >
              <ScrollView
                data={
                  Array [
                    "a",
                    "b",
                    "c",
                  ]
                }
                disableVirtualization={false}
                horizontal={false}
                initialNumToRender={10}
                maxToRenderPerBatch={10}
                numColumns={1}
                onEndReachedThreshold={2}
                removeClippedSubviews={false}
                scrollEventThrottle={50}
                showsVerticalScrollIndicator={false}
                stickyHeaderIndices={Array []}
                updateCellsBatchingPeriod={50}
                viewabilityConfigCallbackPairs={Array []}
                windowSize={21}
              >
                <View
                  style={null}
                >
                  <TouchableOpacity
                    activeOpacity={0.2}
                    style={
                      Object {
                        "padding": 2,
                      }
                    }
                  >
                    <View>
                      <Text>
                        a
                      </Text>
                    </View>
                  </TouchableOpacity>
                </View>
                <View
                  style={null}
                >
                  <TouchableOpacity
                    activeOpacity={0.2}
                    style={
                      Object {
                        "padding": 2,
                      }
                    }
                  >
                    <View>
                      <Text>
                        b
                      </Text>
                    </View>
                  </TouchableOpacity>
                </View>
                <View
                  style={null}
                >
                  <TouchableOpacity
                    activeOpacity={0.2}
                    style={
                      Object {
                        "padding": 2,
                      }
                    }
                  >
                    <View>
                      <Text>
                        c
                      </Text>
                    </View>
                  </TouchableOpacity>
                </View>
              </ScrollView>
            </View>
          </SafeAreaView>
        </View>
      </View>
    </Modal>
  </View>
</View>
`;

exports[`Component: SelectionListBottomSheet Should render a SelectionListBottomSheet with height set 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <Modal
      animationType="none"
      hardwareAccelerated={false}
      supportedOrientations={
        Array [
          "portrait",
          "portrait-upside-down",
          "landscape",
          "landscape-left",
          "landscape-right",
        ]
      }
      transparent={true}
      visible={false}
    >
      <View
        style={
          Array [
            Array [
              Object {
                "backgroundColor": "#00000077",
                "flex": 1,
              },
              undefined,
            ],
            Object {
              "paddingBottom": 0,
            },
          ]
        }
      >
        <TouchableOpacity
          activeOpacity={1}
          style={
            Object {
              "backgroundColor": "transparent",
              "flex": 1,
            }
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "#fff",
              "borderTopLeftRadius": 20,
              "borderTopRightRadius": 20,
              "height": 0,
              "overflow": "hidden",
              "transform": Array [
                Object {
                  "translateX": 0,
                },
                Object {
                  "translateY": 0,
                },
              ],
              "width": "100%",
            }
          }
        >
          <SafeAreaView
            style={
              Object {
                "flex": 1,
              }
            }
          >
            <View
              accessibilityLabel="view-"
              accessible={true}
              style={
                Array [
                  Object {
                    "backgroundColor": "#FFFFFF",
                    "shadowColor": "rgba(0, 0, 0, 0.12)",
                    "shadowOffset": Object {
                      "height": 1,
                      "width": 0,
                    },
                    "shadowRadius": 5,
                  },
                  Object {
                    "shadowOpacity": 0,
                  },
                  Array [
                    Object {
                      "overflow": "visible",
                      "zIndex": 2,
                    },
                    Object {
                      "shadowOpacity": 0,
                    },
                  ],
                ]
              }
              testID="view-"
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "paddingHorizontal": 16,
                    "paddingTop": 12,
                  }
                }
              >
                <View
                  style={
                    Object {
                      "flex": 1,
                    }
                  }
                >
                  <Text
                    accessibilityLabel="txt-"
                    accessible={true}
                    style={
                      Array [
                        Object {
                          "color": "#000000",
                          "fontFamily": "System",
                          "fontWeight": "400",
                          "textAlign": "left",
                        },
                        Array [
                          Object {
                            "color": "#000000",
                            "fontSize": 12,
                          },
                          undefined,
                          undefined,
                          undefined,
                          undefined,
                          Object {
                            "fontFamily": "OpenSans",
                            "fontSize": 16,
                            "lineHeight": 19,
                          },
                          undefined,
                          undefined,
                          undefined,
                          undefined,
                          false,
                          undefined,
                          undefined,
                          false,
                          false,
                          undefined,
                          undefined,
                          undefined,
                          undefined,
                          undefined,
                          undefined,
                          undefined,
                          undefined,
                          false,
                          undefined,
                          undefined,
                          undefined,
                          undefined,
                          undefined,
                          undefined,
                          undefined,
                        ],
                      ]
                    }
                    testID="txt-"
                  >
                    
                  </Text>
                </View>
                <TouchableOpacity
                  accessibilityLabel="btn-close-button"
                  accessible={true}
                  activeOpacity={0.2}
                  style={
                    Object {
                      "alignItems": "flex-end",
                      "height": 44,
                      "justifyContent": "center",
                      "marginRight": 0,
                      "width": 44,
                    }
                  }
                  testID="btn-close-button"
                >
                  <Text
                    allowFontScaling={false}
                    style={
                      Array [
                        Object {
                          "color": undefined,
                          "fontSize": 24,
                        },
                        undefined,
                        Object {
                          "fontFamily": "Material Icons",
                          "fontStyle": "normal",
                          "fontWeight": "normal",
                        },
                        Object {},
                      ]
                    }
                  >
                    Óóç
                  </Text>
                </TouchableOpacity>
              </View>
            </View>
            <View
              style={
                Object {
                  "flex": 1,
                  "paddingHorizontal": 16,
                }
              }
            >
              <ScrollView
                data={
                  Array [
                    "a",
                    "b",
                    "c",
                  ]
                }
                disableVirtualization={false}
                horizontal={false}
                initialNumToRender={10}
                maxToRenderPerBatch={10}
                numColumns={1}
                onEndReachedThreshold={2}
                removeClippedSubviews={false}
                scrollEventThrottle={50}
                showsVerticalScrollIndicator={false}
                stickyHeaderIndices={Array []}
                updateCellsBatchingPeriod={50}
                viewabilityConfigCallbackPairs={Array []}
                windowSize={21}
              >
                <View
                  style={null}
                >
                  <TouchableOpacity
                    activeOpacity={0.2}
                    style={
                      Object {
                        "padding": 2,
                      }
                    }
                  >
                    <View>
                      <Text>
                        a
                      </Text>
                    </View>
                  </TouchableOpacity>
                </View>
                <View
                  style={null}
                >
                  <TouchableOpacity
                    activeOpacity={0.2}
                    style={
                      Object {
                        "padding": 2,
                      }
                    }
                  >
                    <View>
                      <Text>
                        b
                      </Text>
                    </View>
                  </TouchableOpacity>
                </View>
                <View
                  style={null}
                >
                  <TouchableOpacity
                    activeOpacity={0.2}
                    style={
                      Object {
                        "padding": 2,
                      }
                    }
                  >
                    <View>
                      <Text>
                        c
                      </Text>
                    </View>
                  </TouchableOpacity>
                </View>
              </ScrollView>
            </View>
          </SafeAreaView>
        </View>
      </View>
    </Modal>
  </View>
</View>
`;
